### ISSUE TITLE
Facilities support needs auditing / improved

### ISSUE THREAD

**Issue by Fish-Git** [Fish-Git/hyperion#71]
Our current facilities support needs attention.

I can't recall which ones they were off the top of my head, but I believe I came across more than a few instructions that were blindly executing their functionality based solely on their associated `FEATURE_FOOBAR_FACILITY` being `#defined`, without _**also**_ checking whether the given facility was actually _enabled_ or not via the `FACILITY_ENABLED( FOOBAR, ... )` macro.

This is important to do since we currently support dynamically enabling/disabling a given facility at run-time via our `archlvl enable/disable facility` command.

The `#if defined( FEATURE...)` test at _compile-time_ only determines whether or not the instruction's code should get built (should exist), but does _not_ otherwise determine whether that instruction should actually be _enabled_ (allowed to be _executed_).

If the instruction should not actually "exist" (because the bit for that particular facility as stored by the STFL/STFLE instruction is _off_ (i.e. the facility isn't currently logically _enabled_ because the `archlvl` command was used to purposely _disable_ it)) , then the instruction should Program Check with an Operation Exception ("facility not enabled").

Our `archlvl` command's ability to turn facility bits on and off at will also, as a result, currently allows certain invalid facility bit combinations to occur too. This needs to be fixed as well. (We're probably missing some associated FEATURE combination checks in our `featchk.h` header too.)

Finally, since some facility bits are defined by IBM as pertaining to more than one actual z/Architecture facility (see e.g. facility bits 41 and 49), our `archlvl` command and associated `FACTAB` facility table (as well as our `FACILITY_ENABLED` macro too) should allow specifying the facility by any of those names as well.

_Bottom line: **our facilities implementation needs work!**_

**Comment by Fish-Git** [comment]
Coming along nicely!

But now I've reached Issue #73, which I'm going to have to resolve first...

**Comment by ivan-w** [comment]
Because on a z/Arch machine (prior to a z14) a z/Arch machine is an ESA/390 machine once a system reset occurs. And there are NO damage and no architectural prohibition from having z/Arch features enabled in ESA/390 mode (unless they are marked as z/Arch only features - in which case the STFL/STFLE bit should still be on (since the flag states : on z/Arch mode only)

**Comment by ivan-w** [comment]
The PoP is explicit : Bits in STFL/STFLE may be 1 in ESA/390 even if they are not available in ESA/390 mode, since the meaning of the bit is SPECIFIC : The feature is only available in z/Arch mode !

**Comment by ivan-w** [comment]
4-101 : A bit is set to one regardless of the current architectural
mode if its meaning is true. A meaning applies
to the current architectural mode unless it is said to
apply to a specific architectural mode

**Comment by Fish-Git** [comment]
Ivan,

z/Arch PoP, page 4-101:

> A bit is set to one regardless of the current architectural mode if its meaning is true. A meaning applies to the current architectural mode _**unless** it is said to apply to a specific architectural mode._

I interpret this to mean that if the description of a given facility bit contains the phrase "in the z/Architecture architectural mode", that that bit doesn't apply if the machine is currently in ESA/390 mode. Yes, it is still stored as a one since it does apply in z/Arch mode and "a bit is set to one regardless of the current architectural mode if its meaning is true" does apply, so it should definitely be store as a one...

_... on a z/Arch machine that is capable of switching between the two modes._

The z/Arch PoP only applies only to _z/Arch_ machines, _not_ to ESA/390 machines.

The _**ESA/390 PoP**_ applies to a ESA/390 machine _not_ the z/Arch PoP, and the _ESA/390 PoP_ clearly states on page 10-92:

> Bits 6-15 and 21-31 are _reserved_ for indication of new facilities. The bits are currently _**stored as zeros**_ but may be stored as ones in the future.

And bits 6 and 7 on ESA/390 machines (ASN and LX Resuse, and STFLE Extended) fall within that range and thus should be stored as zeros (as well as bits 1-3 and 18 too).

So on a true ESA/390 machine, only bits 0, 4-5, 16-17 and 19-20 can possibly be one. Bits 6-15 and 21-31 are _always_ off (stored as zeros) by default, and bits 1-3 and 18 can only be one on a _z/Arch machine_ (that has those facilities installed and is capable of switching to z/Arch mode), but they can _never_ be on on a true ESA/390 machine.

The confusion stems from the fact that a z/Arch machine supports _both_ modes whereas a ESA/390 machine can only support ESA/390 mode. (duh!)

I've always interpreted the meaning of `archlvl z/Arch` as meaning you want to emulate a true z/Arch machine capable of switching between the two modes, whereas `archlvl ESA/390` means you want to emulate a ESA/390 machine that is only capable of running in ESA/390 mode, _period_. If your config says `archlvl esa/390`, then z/Arch is _not_ installed, _regardless_ of whether Hercules was _built_ with support for z/Arch or not.

Or am I mistaken?

**Comment by ivan-w** [comment]
They key phrase is : "A bit is set to one *regardless* of the current architectural mode"

**Comment by ivan-w** [comment]
And there is NO reason to deprive ESA/390 from z/Architecture features if they are not conflicting with ESA/390. It will NOT break anything.

**Comment by ivan-w** [comment]
Remember : There is NOTHING that prohibits in ANY architecture that prohibits any feature or instruction *UNLESS* it is specifically prohibited (like a bit being set to 0 in the PSW or something like that)

**Comment by Fish-Git** [comment]
> They key phrase is : "A bit is set to one regardless of the current architectural mode"
> 
> So if a facility is available in z/Architecture mode - and ONLY in z/Architecture mode - it shoudl STILL be set to 1 when STFL/STFLE is executed in ESA/390 mode !

Yes! I agree! I said as much.

But that only applies to a z machine that is capable of _both_ modes.

A non-Z machine (pure ESA/390 machine) that _cannot_ run in z/Arch mode, then the bit doesn't apply and is set to zero, as the ESA/390 PoP says.

> And there is NO reason to deprive ESA/390 from z/Architecture features if they are not conflicting with ESA/390. It will NOT break anything.

Oh, but there IS a reason! Such facilities were not available on such machines!

Your statement may be true for Z machines capable of either mode, such when such machines happen to be in ESA/390 mode then there's no reason to deprive ESA/390 _MODE_ on that Z machine from taking advantage of an installed facility, but I'm not talking about Z machines. I'm referring to _NON_-z machines, i.e. machines that can _only_ run in ESA/390 mode. Such machines dis not have such facilities installed on them. So even though our emulator was built with support for that facility, it shouldn't be made available is the archlvl is set to esa/390.

If the archlvl is set to z/Arch mode and the machine just so happens to be in esa/390 mode (due to an IPL or purposely SIGPing into esa/390 mode for example), then yes, there's not reason to deny the machine from taking full advantage to whatever facilities it has installed.

But again, I'm not talking about z machines. I'm talking about ESA/390 machines.

> Remember : There is NOTHING that prohibits in ANY architecture that prohibits any feature or instruction UNLESS it is specifically prohibited (like a bit being set to 0 in the PSW or something like that)

Which is pretty much exactly my point! As I pointed out the _ESA/390_ Pop (_not_ he z/Arch PoP) clearly states:

> Bits 6-15 and 21-31 are reserved for indication of new facilities. The bits are currently _**stored as zeros**_ but may be stored as ones in the future.

"Stored as zeros" means to me the facilities assigned to those bits _**are not available**_.

This means, for example, attempting to execute a STFLE instruction on such a machine should program check because the facility is _not available_.

**Comment by ivan-w** [comment]
Fish,

STFL is only available on ESA/390 systems which are also z/Arch capable. (Per SA22-7201-08 page 10-92)

A bit set to 0 in the facility list does NOT indicate a facility is not installed - It only works the other way around - Basic boolean logic. if a true proposition implies something is true is NOT the same thing as a false proposition implies something is false.
Per wording : It clearly states - a "bit set to 1 indicates"...
Also clarified in z/Arch POP : page 4-101 : 

"STORE FACILITY LIST and STORE FACILITY LIST
EXTENDED may report the absence of a facility,
even though the execution of the facility’s instructions
may appear to indicate that the facility is installed."

**Comment by Fish-Git** [comment]
This issue has already been fixed in my private Bitbucket repository and will eventually appear in my public GitHub repository just as soon as I'm finished with what I'm currently working on.
